### ISSUE TITLE
Add support for Linux

### ISSUE THREAD

**Issue by au2001** [au2001/icloud-passwords-firefox#34]
A custom binary could be created, similar to iCloud Passwords for Windows, which would register a Native messaging host for Firefox and exchange with the iCloud Keychain API.
This would allow to use the extension on Linux.

**Comment by au2001** [comment]
@topiga Sure, any progress would be greatly appreciated!
I guess the authentication part will be fairly similar to what has already been documented on the Hack Different and pypush Discord servers, which I see you're already a part of.
The iCloud Keychain API itself could then be reverse engineered from the iCloud for Windows app, I suppose.

**Comment by topiga** [comment]
Yep, I'll keep you updated on discord if I find anything new. We could also meet in the future if we want to develop an app together, since we are in the same city.

**Comment by janvhs** [comment]
Isn’t it theoretically possible to run iCloud for Windows through Wine or Wine-wrappers like Bottles? You wouldn’t need to reverse engineer the API then.

> I guess the authentication part will be fairly similar to what has already been documented on the Hack Different and pypush Discord servers, which I see you're already a part of.

Is this available somewhere else, too? From experience, discord isn’t the best place to look for documentation.

**Comment by topiga** [comment]
This discord is. Also, I tried running iCloud on wine, but the version with passwords is only available through the Microsoft store, and I don’t know how/don’t want to do that.Le 7 avr. 2024 à 11:47, Jan Fooken ***@***.***> a écrit :﻿
Isn’t it theoretically possible to run iCloud for Windows through Wine or Wine-wrappers like Bottles? You wouldn’t need to reverse engineer the API then.

I guess the authentication part will be fairly similar to what has already been documented on the Hack Different and pypush Discord servers, which I see you're already a part of.

Is this available somewhere else, too? From experience, discord isn’t the best place to look for documentation.

**Comment by mkinz** [comment]
Has any progress been made on this? If not i'd like to pick it up because I would find this very useful

**Comment by jacobbrokaw** [comment]
I'm making the switch to Ubuntu and would love to still be able to use iCloud Passwords. +1 to this!

**Comment by Tiagomachado02** [comment]
Same here. Any progress?

**Comment by dreiness** [comment]
I've been using Chrome for years, even on Linux for convenience, but if there were an extension similar to "iCloud passwords" I'd 100% switch to Firefox.  However I fear the hurdles may be too great since iCloud Passwords is from Apple and likely only created to allow cross usability for Windows users.  They have no interest in supporting Linux.

**Comment by nvillabona** [comment]
Any update on this?

**Comment by miedzinski** [comment]
I've tried running iCloud for Windows using Wine, but had little success - it's a UWP application, which isn't well supported by Wine.

That said, you can run a full Windows virtual machine, install iCloud there, and tunnel communication between the host browser and iCloud in the guest VM over TCP. I've been using this setup for a couple of days, and it works surprisingly well.

@au2001 If you're interested, I’d be happy to prepare a PR.

**Comment by mkinz** [comment]
I'd also be interested in seeing your work on this. There's obviously a small group of us interested in solving this problem so let's collect any working solutions so the group can review and iterate on them. 

Another similar idea: docker image of windows or osx with iCloud installed and similar networking setup. I'm going to try something like this and will report back how it goes.

Thanks

**Comment by miedzinski** [comment]
Glad to hear there's interest in this :) I’ve opened a PR with my setup [here](https://github.com/au2001/icloud-passwords-firefox/pull/61). Would be helpful if someone else could test it and confirm it works for them too. I’ve been using it daily since my last comment without any issues.

**Comment by ghost** [comment]
Is there any movement on this? TIA

**Comment by eljonny** [comment]
I'd be interested to see this happen.

**Comment by kofany** [comment]
me too, waiting on  this 📦

**Comment by LifeWulf** [comment]
I accidentally wiped my Windows install so now I'm experimenting with going Linux full-time; the lack of this add-on is what prevents me from committing fully.

**Comment by 37223** [comment]
+1 for linux support, got blindsided when I did switch from win to linux

**Comment by mkinz** [comment]
So it looks like many people want this, but nobody is actively working on it. I was going to consider taking it on, however there are other password managers that are fully cross platform (i.e. bitwarden) that perform this function quite well. Given open source options that work fine, I wonder if there's a real need to pursue this project?

**Comment by LifeWulf** [comment]
> So it looks like many people want this, but nobody is actively working on it. I was going to consider taking it on, however there are other password managers that are fully cross platform (i.e. bitwarden) that perform this function quite well. Given open source options that work fine, I wonder if there's a real need to pursue this project?

Bitwarden is great and all, but iCloud Passwords is nicely integrated into iPhones and what most iPhone users probably use. Requiring someone to switch password managers just to use Linux adds yet another layer of friction for those looking to move off of Windows. 

It's been a while since I used Bitwarden, but does it have support for 2FA verification codes? That feature of iCloud Passwords alone saves lots of time.

**Comment by miedzinski** [comment]
It's been about half a year since I submitted my PR. Since then I've been using it daily and still haven't experienced any issues. If you're fine with running a Windows VM just for this extension and building it yourself, I recommend checking out my fork.

**Comment by frbelotto** [comment]
Did anyone find a solution? I've installed the firefox icloud extension, but there is no "icloud app" for linux, so, it does not work.

**Comment by 7ijme** [comment]
I'd like to know if anyone is planning on tackling this ;p

**Comment by frbelotto** [comment]
I doubt. I gave uma and moved to Bitwarden, that does work on my iPhone, Linux and windows.

**Comment by Sank6** [comment]
I can take this - I have a prototype working locally. Is there a protocol or format we're using here for the binary to share the password with the extension?

**Comment by mkinz** [comment]
Afaik nobody had touched this in years, and we're all waiting for someone to drive. Looking forward to seeing some progress. I'm happy to help if you need some assistance.

**Comment by Sank6** [comment]
Reference python implementation [here](https://github.com/Sank6/iCloud-Keychain-for-Linux) (with hide my email + a browser extension). This doesn't do much of an attempt of trying to hide from Apple - would be fairly easy to block on Apple's side if they wanted to. Also isn't the best solution:
1. Not a standalone ELF binary
2. Needs an anisette server
3. Uses GNOME's keyring to store the encryption key for passwords - need to modify for non-GNOME Linux setups. 

I think we can fold these python scripts + an anisette server into a standalone executable fairly easily with Rust or something. I don't know enough about Linux to know any setup-agnostic way of storing the encryption key.
